### PR TITLE
[WIP] Add state threading for basic parser

### PR DESCRIPTION
**Pull request by dminuoso**
Adds state threading to parsers to allow for:
-- Running the parser with ST
-- Running the parser with IO
-- Embedding an ST parser in a pure parser
-- Unsafely embedding an IO parser in a pure parser

This is currently work in progress

### CODE REVIEW

**Comment by dminuoso** [comment]
@AndrasKovacs This is a rough initial sketch. This is still missing a Stateful variant, some exports and some haddock.

Can you give a brief look whether this is heading in the right direction?

File: src/FlatParse/Basic.hs [hunk 1]
```diff
@@ -200,71 +208,85 @@ import FlatParse.Internal.UnboxedNumerics
 --------------------------------------------------------------------------------
```
Comment (dminuoso) [inline src/FlatParse/Basic.hs:209]: We introduce a separate type alias for the outer state token passing to make it easier to just thread the state token if you dont care about the actual result type.

**Comment by dminuoso** [comment]
Furthermore, after some talks with Andreas Kleibinger, GHC has `-fstate-hack` which by default is on and the reason IO can compile so efficiently. What this does, is treat any function taking `State# something` as oneShot.

That gives me motivation to drop the entire `proxy#` idea and make parsers polymorphic over `State# s` rather than `k :: ZeroBitType`.

**Comment by AndrasKovacs** [comment]
I find it unlikely that the state hack helps performance. I would be much less surprised to see that it *hurts* performance, and in fact if anything hurts performance when switching to `State#` threading, this would be it. 

Actually, it's good that you mention the state hack, because I was not thinking about it earlier when I wrote that I expect no performance change with `State#`.  Some thoughts:

- Marking functions as oneShot is no good; it loses optimization information. Treating lambdas which are *not* oneShot as such can cause performance bugs. For example, we don't want to inline some independent computation into `chainr`. 
- In most parser code, probably we won't often get such problems, because `Parser` code is forced to be sequential by state-passing (not just `State#`, but actual `Addr#` state!). Problems might occur when computation-heavy *non-Parser* code is embedded into `Parser`. 
- I have spent a fair amount of time looking at Core output for benchmarks and also for larger non-trivial parsers, and it seems generally okay and I don't see room for major improvement. We have definitely more opportunities to lose performance than to gain.

**Comment by dminuoso** [comment]
Honestly I have a really hard time figuring out how the state hack works exactly. I was just going by some rough opinions found on GHC gitlab. There's a lack of good notes inside GHC about how it exactly works.

Either way, we can just observe the diff between `-fno-state-hack` and nothing, if it affects us we have a simple way out.

What is your opinion on `proxy#`? Looking back on how unsafeDupablePerformIO was defined before `runRW#`, it seems marking it as NOINLINE should be safe. If we want that inlineable, we do not have an option other than runRW# and not use `Proxy# ()`

Given that running a Parser likely does not occur in a hotspot, I would be fine with marking it NOINLINE.

**Comment by AndrasKovacs** [comment]
My understanding of `Proxy#` is that it's safe to do whatever with it and we can inline as we like. 

I also vote for just doing IO with `State#` for now.

**Comment by dminuoso** [comment]
Let me illuminate why we have a potential for evil here:

## Background on accursedUnutterablePerformIO
The core idea of `accursedUnutterablePerformIO` is to directly invoke IO by applying the underlying State# threading function to `realWorld#`

```
accursedUnutterablePerformIO (IO m) = case m realWorld# of (# _, r #) -> r
```
By doing it the simplifier can set itself up to optimize out the State# away, which promotes a bunch of other simplifier routines. Consider something like

```haskell
thing = do
  m <- mallocBytes 4096
  n <- mallocBytes 4096
  r <- newIORef 1
  t <- newIORef 1
  ....

f = accursedUnutterablePerformIO thing
```

which is internally equivalent to

```haskell
thing = \state1 -> 
  let (state2, m) = unIO (mallocBytes 4096) state1 in
  let (state3, n) =  unIO (mallocBytes 4096) state2 in
  let (state4, r) = unIO (newIORef 1) state3 in
  let (state5, t) = unIO (newIORef 1) state4 in
  ...

f = thing realWorld#
```

Which the simplifier has ability and authority to inline and turn into
```haskell
f = let m = unIO (mallocBytes 4096) realWorld# in
     let n = m in
     let r = unIO (newIORef 1) realWorld# in
     let t = r in
     ...
```

Yuck! Something akin to the mallocBytes aliasing was the source of numerous memory corruptions and bugs in bytestring. 

https://gitlab.haskell.org/ghc/ghc/-/issues/3486
https://gitlab.haskell.org/ghc/ghc/-/issues/3487
https://gitlab.haskell.org/ghc/ghc/-/issues/7270

What is worse, is that these bugs are hard to trigger because the mood of the simplifier depends on on the surrounding code. Incidentally `accursedUnutterablePerformIO` was previously called `inlinePerformIO`, because back then only difference to unsafeDupablePerformIO was that the latter had a NOINLINE attached to it, which would have prevented exactly this from happening.

## How does this affect us?

a) embedSTinPure and unsafeEmbedIOInPure

If you write code like:

```haskell
f = flip runParser buf $ do
  r <-  unsafeEmbedIOInPure (do m <- mallocBytes 4096
                                                     n <- mallocBytes 4096
                                                     r <- newIORef 1
                                                     t <- newIORef 1
                                               ...)
  ...

g = flip runParser buf $ do
  r <-  embedSTInPure (do r <- newSTRef 1
                                           t <- newSTRef 1
                                     ...)
  ...
```

and runParser is allowed to inline, then the proxy# token can be floated into the IO and ST action. Accursed and utterable. Yuck again!

Another source of problems is our new `byteString`. In your original adaption there was no correct threading of the `touch#` token, which I addressed by using `reallyUnsafeStateCoerce`.

```haskell
byteString :: B.ByteString -> ParserT st e ()
byteString (B.PS (ForeignPtr bs fcontent) _ (I# len)) =
  ...
  -- We roundtrip through ParserIO in order to use touch#
  in reallyUnsafeStateCoerce $
       ParserT \fp eob s rw -> case len <=# minusAddr# eob s of
         1# -> case go64 bs (plusAddr# bs len) s rw of
                 (# rw', res #) -> case touch# fcontent rw' of
                   rw'' ->```
```
If inlining on both `runParser` and `byteString` is allowed to occur, the state token threading can be optimized away entirely, and `touch#` can be optimized away entirely. Yuck once again!

For very similar reasons the `noDuplicate#` in `unsafeEmbedIOInPure` can be optimized away if the state token gets inlined, magically turning a use of `unsafeEmbedIOInPure` into `unsafeDupableEmbedIOInPure`. Yuck!

## Conclusion
 
Without some extra care, `embedSTinPure` and `unsafeEmbedIOInPure` becomes accursed and unutterable, and we lack the ability of safely using touch# as part of a parser, unless we do to extra lengths of using unsafePerformIO (the purpose of which here would be only to prevent the state token from being inlined).

But this is all easily avoided if we just mark runParser as NOINLINE.

Further I would really like to keep the ability to run IO inside a parser.

My original motivation behind writing all of this, was that I be able to:

```haskell
f = do
  w <- word8
#ifdef DEBUG_ENABLED
  unsafeLiftIO (putStrLn ("tag is: " <> show w))
#endif
```

Many of my earlier attempts of using unsafePerformIO to get the same thwarted by the simplifier, with some of these debugging hooks being optimized entirely or reordered quite senselessly. Even some of my attempts at making `tracing :: (a -> Text) -> Parser e a -> Parser e a` were thwarted by the optimizer. There's something to be said of the powers of the simplifier.

I would be surprised if others wouldn't want easy access to debugging inside their parser as well, but I don't have strong feelings about it since I can just define unsafeEmbedIOInPure and unsafeLiftIO myself in my programs. But it is not something we should ask users for.

Given that as long as we keep runParser as NOINLINE, then `unsafeDupablePerformIO / unsafeEmbedIOInPure` is as safe as unsafePerformIO, and it's such a low hanging fruit that I do not see the advantage of not providing it.

**Comment by AndrasKovacs** [comment]
I know about accursed running. To clarify my stance:
- The only way a function taking `State#` argument should be run is by `runRW#`. 
- We should ideally never cast between different state types or `State#` and `Proxy#`. Although there is unsafe coercion between `ST` and `IO` in `base` (`unsafeSTToIO`, `unsafeIOToST`), I don't think such conversion is practically very useful. In `IO` we can't do `ST` actions and in `ST` we can't do `IO` actions, so we should instead just `runRW#` `IO` in `ST` or `runST` `ST` in `IO`, without unsafe coercion.
- Likewise, there should be no casting between state types in `bytestring`.  Instead, the implementation is simply an `IO` action embedded in any `Parser` by way of `runRW`. Then, the `touch#` operates on `State# RealWorld` so there's no issue.

**Comment by dminuoso** [comment]
We should then use unsafeDupablePerformIO rather than runRW# or use it in conjunction with lazy. But I think unsafeDupablePerformIO is safer in that we really want its semantics, but it will give us better long term compatibility in case other unknown or future bugs like the following creep up:

https://gitlab.haskell.org/ghc/ghc/-/issues/19181 
https://gitlab.haskell.org/ghc/ghc/-/issues/19413

Anyway, I really do want some kind of (safe) `unsafeLiftIO :: IO a -> Parser e a`. We can't safely conjure up `unsafeLiftIO = pure . unsafePerformIO` because with proxy# floating in, we do not have any guarantee about effect ordering, or whether the effect is ran multiple times - exactly the type of problem why I started this effort in the first place. Switching an entire parser to ParserIO just to be able to drop the occasional debug print, or have IO effects to only run if the flag DEBUG_ENABLE is passed, is not something I want to do either.

Also, it does seem useful to have a parser section that relies on poking into mutable buffers like `embedSTinPure` for the same reasons runST is useful.

I feel like we'd giving up these nice features on the face of slightly improved elegance of Proxy#

**Comment by AndrasKovacs** [comment]
Hm, so you say that you want to be able to run IO in pure parsers with exactly the same ordering guarantees as when running IO in IO? And that you don't like the default robustness of using `unsafePerformIO` (or any of the tracing functions from `Debug.Trace`)? 

- Indeed, the various `unsafePerforms` don't have the same guarantees as doing proper `IO`, but personally I have not had issues with using `traceM` for tracing in various monads, including the `Parser` in this library. Maybe the reason is that I generally only trace with `-O0`, and you'd like to trace with optimization as well?
- Could we make IO embedding more robust by introducing a black-box dependency on the actual parsing state? For example, by having a `NOINLINE` `embedIO` function which takes and returns an extra `Addr#?`. 

I still do not quite understand why applying `proxy#` would ever be dangerous or problematic. I could replace it by any type and value, for the purpose of passing an unused state.

**Comment by dminuoso** [comment]
I think `proxy#` is a bit of a red herring at this point and detrimental to the crux of the discussion.

Let's assume a

```haskell
unsafeLiftIO :: IO a -> Parser e a
unsafeLiftIO io = pure (unsafePerformIO io)
{-# NOINLINE unsafeLiftIO #-}
```

If we inline `pure`, we get
```haskell
unsafeLiftIO io = let a = unsafePerformIO io
                  in ParserT \fp eob s st -> OK# st a s
```

The action is still deferred until such a time that `a` is forced. Given that we're already very sequenced logic, let's see if we can address that. We can `seq` the action before producing OK

```haskell
unsafeLiftIO io = let !a = unsafePerformIO io
                  in ParserT \fp eob s st -> OK# st a s
```

Note, we already have the Addr# threading by virtue of ParserT. Now since unsafeLiftIO is forbidden to be inlined, it cannot float out. Because of bang, it will be forced before OK is produced. Maybe this is fine the way it is, in the presence of optimizations I think this would work out just fine. Do you agree?

As for the other parts:

There is also the idea of writing some ParserST code in one module, and run it in Parser code, similar to how you might use runST in pure code. The inverse is possible at least, given that a `Parser` can be replaced with `ParserT` (since ParserT subsumes Parser). 

Proposition:

Let us toss out these state token coercions (`ParserST->ParserIO`, `ParserIO->ParserST`, `ParserIO->Parser`, `ParserST->Parser`), but keep `{-# NOINLINE runParser #-}`. That way the odd power user like myself can just conjure up reallyUnsafeStateCoerce and do what he or she wants. For the more likely debugging usecase we can just provide `unsafeLiftIO` as defined above for convenience.

**Comment by AndrasKovacs** [comment]
> Do you agree?

I agree. 

I agree to your proposition as well.

**Comment by dminuoso** [comment]
This is ready for a final review now.

**Pull request merged**